api/src/services/authorization.js

```diff
@@ -559,90 +585,120 @@ const isSensitive = (userCtx, subject, submitter, isPrivate, allowedSubmitter) =
   if (!usesReportDepth(authorizationContext)) {
     // no depth limitation
     return true;
   }
 
-  const [{ value: { type: docType } = {} }] = replicationKeys;
-  if (docType !== 'data_record') {
+  if (docByReplicationKey.type !== 'data_record') {
     // allow everything that's not a data_record through (f.e. targets)
     return true;
   }
 
-  return replicationKeys.some(replicationKey => {
-    const { key: subject, value: { submitter } = {} } = replicationKey;
-    if (submitter === authorizationContext.userCtx.contact_id) {
-      // current user is the submitter
-      return true;
-    }
-
-    return authorizationContext.subjectsDepth[subject] <= authorizationContext.reportDepth;
-  });
-};
-
-const groupViewResultsById = (authorizationContext, viewResults) => {
-  if (!usesReportDepth(authorizationContext)) {
-    return {};
+  if (docByReplicationKey.submitter === authorizationContext.userCtx.contact_id) {
+    // current user is the submitter
+    return true;
   }
 
-  return _.groupBy(viewResults.rows, 'id');
-};
+  const replicationKeys = Array.isArray(docByReplicationKey.key) ?
+    docByReplicationKey.key : [docByReplicationKey.key];
 
-const isTaskDoc = (row) => row.value && row.value.type === 'task';
+  return replicationKeys.some(replicationKey => {
+    return authorizationContext.subjectsDepth[replicationKey] <= authorizationContext.reportDepth;
+  });
+};
 
 // casting everything to string to insure that comparisons are against the same type to avoid 5 > 'a' || 5 < 'a'
 const prepareForSortedSearch = array => array.map(element => String(element)).sort();
 const sortedIncludes = (sortedArray, element) => _.sortedIndexOf(sortedArray, String(element)) !== -1;
 
+
 /**
  * Returns a list of document ids that the user is allowed to see and edit
- * @param {AuthorizationContext} authorizationContext
- * @returns {Promise<string[]>}
+ * @param authorizationContext
+ * @returns {Promise<{ id: string, fields: DocByReplicationKey }[]>}
  */
-const getDocsByReplicationKey = (authorizationContext) => {
-  return db.medic.query('medic/docs_by_replication_key', { keys: authorizationContext.subjectIds }).then(results => {
-    const viewResultsById = groupViewResultsById(authorizationContext, results);
+const getDocsByReplicationKeyNouveau = async (authorizationContext) => {
+  const allKeys = [...authorizationContext.subjectIds];
+  const hits = [];
+
+  while (allKeys.length) {
+    const chunk = allKeys.splice(0, nouveau.BATCH_LIMIT);
+    const response = await request.post({
+      uri: `${environment.couchUrl}/_design/medic/_nouveau/docs_by_replication_key`,
+      body: {
+        q: `key:(${chunk.map(nouveau.escapeKeys).join(' OR ')})`,
+        limit: nouveau.RESULTS_LIMIT,
+      }
+    });
+
+    if (!response.hits || !response.hits.length) {
+      continue;
+    }
+
+    hits.push(...response.hits.map(hit => ({
+      id: hit.id,
+      fields: hit.fields,
```

Comment:
Minor, but if we massaged `fields.key` here to always be an array, we would not have to have that logic repeated several places above...

Reply:
I agree, but I haven't found the best way to get that data to be consistent.